outlines/processors/base_logits_processor.py

```diff
@@ -29,50 +30,100 @@ class BaseLogitsProcessor(Protocol):
     ) -> torch.Tensor:
-        ...
+        """
+        input_ids and logits are always 2D tensors for handling a batch of sequences.
+
+        - input_ids.shape[1] -> contains the sequence of int tokens
+        - logits.shape[0] -> Dimension 1 contains one
+
+        Important to keep in mind when designing universal logits processors
+        - logits processors are only used once and never re-applied for a new sequence generator
+        - Some models only pass output_ids, some models such as llamacpp and transformers prefix with input_ids
+        - Some sampling methods, such as beam search, result in unstable sequence ordering in models like vLLM
+        """
+        pass
 
+    @torch.no_grad()
     def __call__(
         self,
         input_ids: Union[NDArray[np.int64], List[int], torch.Tensor],
         logits: Union[NDArray[np.float32], torch.Tensor],
     ) -> Union[NDArray[np.int64], torch.Tensor]:
         """
         Apply logits processor
+
         Unify type
-        - convert input_ids: either ndarray, List[int], or Tensor -> List[int]
-        - convert logits: either ndarray, mlx array, Tensor -> Tensor
+        - convert input_ids: either ndarray, List[int], or Tensor -> 2D tensor
+        - convert logits: either ndarray, mlx array, Tensor -> 2D Tensor
+
         Call process_logits() to perform business logic
         """
-        with torch.no_grad():
-            if not isinstance(input_ids, list):
-                input_ids = input_ids.tolist()
-
-            if isinstance(logits, np.ndarray):
-                # Unify type, convert numpy array to Tensor
-                # from_numpy and .numpy() don't copy the data, it uses the same memory address
-                torch_logits = torch.from_numpy(logits)
-                processed_torch_logits = self.process_logits(input_ids, torch_logits)
-                return processed_torch_logits.detach().numpy()
-
-            elif isinstance(logits, torch.Tensor):
-                return self.process_logits(input_ids, logits)
-
-            elif is_mlx_array(logits):
-                # mlx -> torch -> mlx conversion docs:
-                # https://ml-explore.github.io/mlx/build/html/usage/numpy.html
-                import mlx.core as mx
-
-                torch_logits = torch.from_dlpack(logits)
-                processed_torch_logits = self.process_logits(input_ids, torch_logits)
-
-                # numpy doesn't support bfloat16, mlx doesn't support direct conversion from torch
-                logits_float32_numpy = processed_torch_logits.float().numpy()
-                return mx.array(logits_float32_numpy)
-
-            else:
-                raise TypeError(
-                    "LogitsProcessor must be called with either np.NDArray"
-                    ", torch.Tensor, or mlx.core.array typed logits"
-                )
+
+        # ensure logits are torch Tensors
+        torch_logits = self._to_torch(logits)
```

Comment:
What is the cost of this conversion? We need to profile this.

Reply:
The idea is that we can cast `numpy` and `mlx.core` arrays to `torch.Tensor` using shared memory, **not** a copy operation :)

Then we can implement a single `OutlinesLogitsProcessor` class with a single `process_logits()` method and it works out of the box for any library. While not in `main` yet, I've tested `FSMLogitsProcessor`, and `RegexLogitsProcessor` with every nearly `outlines.models` option and they work with no additional changes.

This makes the processors easy to maintain and implement.